File: chart/src/main/java/cn/jingzhuan/lib/chart3/base/AbstractChartView.java
package cn.jingzhuan.lib.chart3.base;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.util.AttributeSet;
import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import cn.jingzhuan.lib.chart.R;
import cn.jingzhuan.lib.chart.component.Axis;
import cn.jingzhuan.lib.chart.component.AxisX;
import cn.jingzhuan.lib.chart.component.AxisY;
import cn.jingzhuan.lib.chart.component.Highlight;
import cn.jingzhuan.lib.chart.data.AbstractDataSet;
import cn.jingzhuan.lib.chart.data.ChartData;
import cn.jingzhuan.lib.chart.data.Value;
import cn.jingzhuan.lib.chart3.renderer.AxisRenderer;
import cn.jingzhuan.lib.chart3.renderer.HighlightRenderer;
import cn.jingzhuan.lib.chart3.state.HighlightState;

/**
 * @since 2023-09-01
 * created by lei
 */
public abstract class AbstractChartView<V extends Value, T extends AbstractDataSet<V>> extends ScrollAndScaleView implements IChartView{

    protected WeakReference<Bitmap> mDrawBitmap;

    protected Canvas mBitmapCanvas;

    protected Bitmap.Config mBitmapConfig = Bitmap.Config.ARGB_8888;

    protected AxisRenderer<V, T> mAxisLeftRenderer;

    protected AxisRenderer<V, T> mAxisRightRenderer;

    protected AxisRenderer<V, T> mAxisTopRenderer;

    protected AxisRenderer<V, T> mAxisBottomRenderer;

    protected HighlightRenderer<V, T> mHighlightRenderer;

    private int minChartWidth;

    private int minChartHeight;

    /**
     * 背景颜色
     */
    private int backgroundColor;

    /**
     * 坐标轴刻度文本 是否画在底层
     */
    private boolean drawLabelsInBottom;

    /**
     * 是否需要展示水印
     */
    private boolean showWaterMark;

    /**
     * 是否黑夜模式
     */
    private boolean isNightMode;

    public AbstractChartView(Context context) {
        super(context);
    }

    public AbstractChartView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public AbstractChartView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public AbstractChartView(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void init(AttributeSet attrs, int defStyleAttr) {
        super.init(attrs, defStyleAttr);
        TypedArray ta = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.Chart, defStyleAttr, defStyleAttr);

        try {
            int minChartWidth = ta.getDimensionPixelSize(R.styleable.Chart_minChartWidth, 0);
            setMinChartWidth(minChartWidth);

            int minChartHeight = ta.getDimensionPixelSize(R.styleable.Chart_minChartHeight, 0);
            setMinChartHeight(minChartHeight);

            int backgroundColor = ta.getColor(R.styleable.Chart_backgroundColor, Color.TRANSPARENT);
            setBackgroundColor(backgroundColor);

            boolean drawLabelsInBottom = ta.getBoolean(R.styleable.Chart_drawLabelsInBottom, false);
            setDrawLabelsInBottom(drawLabelsInBottom);

            boolean showWaterMark = ta.getBoolean(R.styleable.Chart_showWaterMark, false);
            setShowWaterMark(showWaterMark);

            boolean isNightMode = ta.getBoolean(R.styleable.Chart_isNightMode, false);
            setNightMode(isNightMode);

            initAxisRenderers(ta);

            mHighlightRenderer = new HighlightRenderer<>(this);

            ta.recycle();
        } catch (Exception e) {
            ta.recycle();
        }

        initChart();

    }

    private void initAxisRenderers(TypedArray ta) {
        AxisY axisLeft = new AxisY(AxisY.LEFT_INSIDE);
        mAxisLeftRenderer = new AxisRenderer<>(this, axisLeft);

        AxisY axisRight = new AxisY(AxisY.RIGHT_INSIDE);
        mAxisRightRenderer = new AxisRenderer<>(this, axisRight);

        AxisX axisTop = new AxisX(AxisX.TOP);
        mAxisTopRenderer = new AxisRenderer<>(this, axisTop);

        AxisX axisBottom = new AxisX(AxisX.BOTTOM);
        mAxisBottomRenderer = new AxisRenderer<>(this, axisBottom);

        List<AxisRenderer<V, T>> mAxisRenderers = new ArrayList<>(4);
        mAxisRenderers.add(mAxisLeftRenderer);
        mAxisRenderers.add(mAxisRightRenderer);
        mAxisRenderers.add(mAxisTopRenderer);
        mAxisRenderers.add(mAxisBottomRenderer);

        float labelTextSize = ta.getDimension(R.styleable.Chart_labelTextSize, 28);

        float labelSeparation = ta.getDimensionPixelSize(R.styleable.Chart_labelSeparation, 10);

        int labelTextColor = ta.getColor(R.styleable.Chart_labelTextColor, Color.GRAY);

        float gridThickness = ta.getDimension(R.styleable.Chart_gridThickness, 2);

        float axisThickness = ta.getDimension(R.styleable.Chart_axisThickness, 2);

        int gridColor = ta.getColor(R.styleable.Chart_gridColor, Color.GRAY);

        int axisColor = ta.getColor(R.styleable.Chart_axisColor, Color.GRAY);

        for (AxisRenderer<V, T> axisRenderer : mAxisRenderers) {
            Axis axis = axisRenderer.getAxis();
            axis.setLabelTextSize(labelTextSize);
            axis.setLabelTextColor(labelTextColor);
            axis.setLabelSeparation(labelSeparation);
            axis.setGridColor(gridColor);
            axis.setGridThickness(gridThickness);
            axis.setAxisColor(axisColor);
            axis.setAxisThickness(axisThickness);
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        // Clips the next few drawing operations to the content area
        int clipRestoreCount = canvas.save();

        // 剪切ContentRect区域
        canvas.clipRect(mContentRect);

        // 画背景
        if (getBackgroundColor() != Color.TRANSPARENT){
            canvas.drawColor(getBackgroundColor());
        }

        // 画水印
        if (isShowWaterMark()) {
            drawWaterMark(canvas);
        }

        // 画坐标轴
        drawAxis(canvas);

        // 画网格线
        drawGridLine(canvas);

        // 画坐标轴文本 (左、右、上)
        if (isDrawLabelsInBottom()) {
            drawAxisLabels(canvas);
        }

        createBitmapCache();

        if (getBitmapCanvas() != null && getRenderPaint() != null) {

            drawChart(getBitmapCanvas());

            canvas.drawBitmap(getDrawBitmap(), 0, 0, getRenderPaint());
        }

        // Removes clipping rectangle
        canvas.restoreToCount(clipRestoreCount);

        if (!isDrawLabelsInBottom()) {
            drawAxisLabels(canvas);
        }

        if (mAxisBottomRenderer.getAxis().getLabelHeight() != 0) {
            drawBottomLabels(canvas);
        }

        // 画十字光标
        if (getHighlightState() != HighlightState.initial) {
            drawHighlight(canvas);
        }
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldWidth, int oldHeight) {
        Axis axisLeft = getAxisLeftRenderer().getAxis();
        int chartLeft = getPaddingLeft() + (axisLeft.isInside() ? 0 : axisLeft.getLabelWidth());

        Axis axisRight = getAxisRightRenderer().getAxis();
        int chartRight = getWidth() - getPaddingRight() - (axisRight.isInside() ? 0 : axisRight.getLabelWidth());

        Axis axisBottom = getAxisBottomRenderer().getAxis();
        int contentBottom = getHeight() - getPaddingBottom() - axisBottom.getLabelHeight();

        mContentRect.set(chartLeft, getPaddingTop(), chartRight,contentBottom);

        mBottomRect.set(chartLeft, contentBottom, chartRight, getHeight());
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        Axis axisLeft = getAxisLeftRenderer().getAxis();
        int width = getMinChartWidth() + getPaddingLeft() + (axisLeft.isInside() ? 0 : axisLeft.getLabelWidth()) + getPaddingRight();

        Axis axisBottom = getAxisBottomRenderer().getAxis();
        int height = getMinChartHeight() + (axisBottom.isInside() ? 0 : axisBottom.getLabelHeight()) + getPaddingBottom();
        setMeasuredDimension(
                Math.max(getSuggestedMinimumWidth(), resolveSize(width, widthMeasureSpec)),
                Math.max(getSuggestedMinimumHeight(), resolveSize(height, heightMeasureSpec))
        );
    }

    public void setTypeface(Typeface tf) {
        mAxisLeftRenderer.setTypeface(tf);
        mAxisLeftRenderer.setTypeface(tf);
        mAxisTopRenderer.setTypeface(tf);
        mAxisBottomRenderer.setTypeface(tf);
    }

    protected void createBitmapCache() {
        int width = getContentRect().width() + getContentRect().left;
        int height = getContentRect().height();

        if (mDrawBitmap == null
                || (mDrawBitmap.get() == null)
                || (mDrawBitmap.get().getWidth() != width)
                || (mDrawBitmap.get().getHeight() != height)) {

            if (width > 0 && height > 0) {
                mDrawBitmap = new WeakReference<>(Bitmap.createBitmap(getResources().getDisplayMetrics(), width, height, mBitmapConfig));
                mBitmapCanvas = new Canvas(mDrawBitmap.get());
            } else
                return;
        }

        mDrawBitmap.get().eraseColor(Color.TRANSPARENT);
    }

    protected Bitmap getDrawBitmap() {
        return mDrawBitmap.get();
    }

    public void releaseBitmap() {
        if (mBitmapCanvas != null) {
            mBitmapCanvas.setBitmap(null);
            mBitmapCanvas = null;
        }
        if (mDrawBitmap != null) {
            if (mDrawBitmap.get() != null) mDrawBitmap.get().recycle();
            mDrawBitmap.clear();
            mDrawBitmap = null;
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        releaseBitmap();
    }

    public Canvas getBitmapCanvas() {
        return mBitmapCanvas;
    }

    public int getMinChartWidth() {
        return minChartWidth;
    }

    public void setMinChartWidth(int minChartWidth) {
        this.minChartWidth = minChartWidth;
    }

    public int getMinChartHeight() {
        return minChartHeight;
    }

    public void setMinChartHeight(int minChartHeight) {
        this.minChartHeight = minChartHeight;
    }

    public int getBackgroundColor() {
        return backgroundColor;
    }

    public void setBackgroundColor(int backgroundColor) {
        this.backgroundColor = backgroundColor;
    }

    public boolean isDrawLabelsInBottom() {
        return drawLabelsInBottom;
    }

    public void setDrawLabelsInBottom(boolean drawLabelsInBottom) {
        this.drawLabelsInBottom = drawLabelsInBottom;
    }

    public boolean isShowWaterMark() {
        return showWaterMark;
    }

    public void setShowWaterMark(boolean showWaterMark) {
        this.showWaterMark = showWaterMark;
    }

    public boolean isNightMode() {
        return isNightMode;
    }

    public void setNightMode(boolean nightMode) {
        isNightMode = nightMode;
    }

    // <editor-fold desc="十字光标 配置">    ----------------------------------------------------------

    /**
     * 是否需要展示水平交叉线-默认展示
     */
    private boolean enableHorizontalHighlight = true;

    public boolean isEnableHorizontalHighlight() {
        return enableHorizontalHighlight;
    }

    public void setEnableHorizontalHighlight(boolean enableHorizontalHighlight) {
        this.enableHorizontalHighlight = enableHorizontalHighlight;
    }

    /**
     * 是否需要展示垂直交叉线-默认展示
     */
    private boolean enableVerticalHighlight = true;

    public boolean isEnableVerticalHighlight() {
        return enableVerticalHighlight;
    }

    public void setEnableVerticalHighlight(boolean enableVerticalHighlight) {
        this.enableVerticalHighlight = enableVerticalHighlight;
    }

    /**
     * 是否需要展示水平交叉线左边文本-默认展示
     */
    private boolean enableHighlightLeftText = true;

    public boolean isEnableHighlightLeftText() {
        return enableHighlightLeftText;
    }

    public void setEnableHighlightLeftText(boolean enableHighlightLeftText) {
        this.enableHighlightLeftText = enableHighlightLeftText;
    }

    /**
     * 是否需要展示水平交叉线右边文本-默认不展示
     */
    private boolean enableHighlightRightText = false;

    public boolean isEnableHighlightRightText() {
        return enableHighlightRightText;
    }

    public void setEnableHighlightRightText(boolean enableHighlightRightText) {
        this.enableHighlightRightText = enableHighlightRightText;
    }

    /**
     * 是否需要展示垂直交叉线底部文本-默认展示
     */
    private boolean enableHighlightBottomText = true;

    public boolean isEnableHighlightBottomText() {
        return enableHighlightBottomText;
    }

    public void setEnableHighlightBottomText(boolean enableHighlightBottomText) {
        this.enableHighlightBottomText = enableHighlightBottomText;
    }

    /**
     * 交叉线文本大小
     */
    private float mHighlightTextSize;

    public void setHighlightTextSize(float textSize) {
        this.mHighlightTextSize = textSize;
    }

    public float getHighlightTextSize() {
        return this.mHighlightTextSize;
    }

    /**
     * 交叉线文本颜色
     */
    private int mHighlightTextColor = Color.TRANSPARENT;

    public int getHighlightTextColor() {
        return mHighlightTextColor;
    }

    public void setHighlightTextColor(int mHighlightTextColor) {
        this.mHighlightTextColor = mHighlightTextColor;
    }

    /**
     * 交叉线文本背景颜色
     */
    private int mHighlightTextBgColor = Color.TRANSPARENT;

    public int getHighlightTextBgColor() {
        return mHighlightTextBgColor;
    }

    public void setHighlightTextBgColor(int mHighlightTextBgColor) {
        this.mHighlightTextBgColor = mHighlightTextBgColor;
    }

    /**
     * 交叉线文本背景高度
     */
    private int mHighlightTextBgHeight;

    public int getHighlightTextBgHeight() {
        return mHighlightTextBgHeight;
    }

    public void setHighlightTextBgHeight(int mHighlightTextBgHeight) {
        this.mHighlightTextBgHeight = mHighlightTextBgHeight;
    }

    /**
     * 交叉线厚度
     */
    private float mHighlightThickness = 3;

    public float getHighlightThickness() {
        return mHighlightThickness;
    }

    public void setHighlightThickness(float highlightThickness) {
        this.mHighlightThickness = highlightThickness;
    }

    // </editor-fold desc="十字光标 配置">    ----------------------------------------------------------

    public AxisRenderer<V, T> getAxisLeftRenderer() {
        return mAxisLeftRenderer;
    }

    public AxisRenderer<V, T> getAxisRightRenderer() {
        return mAxisRightRenderer;
    }

    public AxisRenderer<V, T> getAxisTopRenderer() {
        return mAxisTopRenderer;
    }

    public AxisRenderer<V, T> getAxisBottomRenderer() {
        return mAxisBottomRenderer;
    }

    public abstract ChartData<T> getChartData();

    abstract Paint getRenderPaint();

    /**
     * 十字光标选中
     */
   abstract void highlightValue(Highlight highlight);

    /**
     * 清除十字光标
     */
    abstract void cleanHighlight();
}

File: chart/src/main/java/cn/jingzhuan/lib/chart3/base/AbstractChartView.kt
package cn.jingzhuan.lib.chart3.base

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import android.util.ArrayMap
import android.util.AttributeSet
import cn.jingzhuan.lib.chart.R
import cn.jingzhuan.lib.chart.animation.ChartAnimator
import cn.jingzhuan.lib.chart3.Highlight
import cn.jingzhuan.lib.chart3.axis.AxisX
import cn.jingzhuan.lib.chart3.axis.AxisY
import cn.jingzhuan.lib.chart3.data.ChartData
import cn.jingzhuan.lib.chart3.data.dataset.AbstractDataSet
import cn.jingzhuan.lib.chart3.renderer.AxisRenderer
import cn.jingzhuan.lib.chart3.renderer.DrawLineRenderer
import cn.jingzhuan.lib.chart3.renderer.HighlightRenderer
import cn.jingzhuan.lib.chart3.renderer.RangeRenderer
import cn.jingzhuan.lib.chart3.utils.ChartConstant.TYPE_AXIS_BOTTOM
import cn.jingzhuan.lib.chart3.utils.ChartConstant.TYPE_AXIS_LEFT
import cn.jingzhuan.lib.chart3.utils.ChartConstant.TYPE_AXIS_RIGHT
import cn.jingzhuan.lib.chart3.utils.ChartConstant.TYPE_AXIS_TOP
import kotlin.math.max

/**
 * @since 2023-09-05
 * created by lei
 */
abstract class AbstractChartView<T : AbstractDataSet<*>> : ScrollAndScaleView, IChartView {

    val axisLeft = AxisY(AxisY.LEFT_INSIDE)

    val axisRight = AxisY(AxisY.RIGHT_INSIDE)

    val axisTop = AxisX(AxisX.TOP)

    val axisBottom = AxisX(AxisX.BOTTOM)

    val axisRenderers: ArrayMap<Int, AxisRenderer<T>> = ArrayMap(4)

    protected lateinit var highlightRenderer: HighlightRenderer<T>

    protected lateinit var rangeRenderer: RangeRenderer<T>

    protected lateinit var drawLineRenderer: DrawLineRenderer<T>

    var minChartWidth = 0

    var minChartHeight = 0

    /**
     * 背景颜色
     */
    var bgColor = 0

    /**
     * 坐标轴刻度文本 是否画在底层
     */
    var isDrawLabelsInBottom = false

    /**
     * 是否需要展示水印
     */
    var isShowWaterMark = false

    /**
     * 是否黑夜模式
     */
    var isNightMode = false

    // <editor-fold desc="十字光标 配置">    ----------------------------------------------------------
    /**
     * 是否需要展示水平交叉线-默认展示
     */
    var isEnableHorizontalHighlight = true

    /**
     * 是否需要展示垂直交叉线-默认展示
     */
    var isEnableVerticalHighlight = true

    /**
     * 是否需要展示水平交叉线左边文本-默认展示
     */
    var isEnableHighlightLeftText = true

    /**
     * 是否需要展示水平交叉线右边文本-默认不展示
     */
    var isEnableHighlightRightText = false

    /**
     * 是否需要展示垂直交叉线底部文本-默认展示
     */
    var isEnableHighlightBottomText = true

    /**
     * 是否需要展示垂直交叉线折线圆点(只有折线有效)-默认不展示
     */
    var isEnableHighlightLineRoundDot = false

    /**
     * 垂直交叉线折线圆点颜色(只有折线有效)
     */
    var highlightLineRoundDotOutColor = 0

    /**
     * 垂直交叉线折线圆点半径(只有折线有效)
     */
    var highlightLineRoundDotOutRadius = 0

    var highlightLineRoundDotInRadius = 0

    /**
     * 交叉线文本大小
     */
    var highlightTextSize = 0

    /**
     * 交叉线文本颜色
     */
    var highlightTextColor = 0

    /**
     * 交叉线文本背景高度
     */
    var highlightTextBgHeight = 0

    /**
     * 交叉线厚度
     */
    var highlightThickness = 0

    /**
     * 交叉线颜色
     */
    var highlightColor = 0

    // </editor-fold desc="十字光标 配置">    ----------------------------------------------------------

    /**
     * 需要保留的小数位
     */
    var decimalDigitsNumber = 2

    /**
     * 是否展示最大最小值
     */
    var isShowMaxMinValue = false

    /**
     * 最大最小值文本大小
     */
    var maxMinValueTextSize = 0

    /**
     * 最大最小值文本颜色
     */
    var maxMinValueTextColor = 0

    var offsetPercent = 0f

    var focusIndex = -1

    /**
     * 是否展示现价线
     */
    var isShowLastPriceLine = false

    /**
     * 现价线颜色
     */
    var lastPriceLineColor = 0

    constructor(context: Context?) : super(context) {
        init(null, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(attrs, defStyleAttr)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(attrs, defStyleAttr)
    }

    final override fun init(attrs: AttributeSet?, defStyleAttr: Int) {
        super.init(attrs, defStyleAttr)
        val ta = context.theme.obtainStyledAttributes(attrs, R.styleable.Chart, defStyleAttr, defStyleAttr)
        try {
            val minChartWidth = ta.getDimensionPixelSize(R.styleable.Chart_minChartWidth, 0)
            this.minChartWidth = minChartWidth

            val minChartHeight = ta.getDimensionPixelSize(R.styleable.Chart_minChartHeight, 0)
            this.minChartHeight = minChartHeight

            val backgroundColor = ta.getColor(R.styleable.Chart_backgroundColor, Color.TRANSPARENT)
            this.bgColor = backgroundColor

            val drawLabelsInBottom = ta.getBoolean(R.styleable.Chart_drawLabelsInBottom, false)
            this.isDrawLabelsInBottom = drawLabelsInBottom

            val showWaterMark = ta.getBoolean(R.styleable.Chart_showWaterMark, false)
            isShowWaterMark = showWaterMark

            val isNightMode = ta.getBoolean(R.styleable.Chart_isNightMode, false)
            this.isNightMode = isNightMode

            val highlightColor = ta.getColor(R.styleable.Chart_highlightColor, Color.BLACK)
            this.highlightColor = highlightColor

            val highlightThickness = ta.getDimensionPixelSize(R.styleable.Chart_highlightThickness, 3)
            this.highlightThickness = highlightThickness

            val highlightTextBgHeight = ta.getDimensionPixelSize(R.styleable.Chart_highlightTextBgHeight, 50)
            this.highlightTextBgHeight = highlightTextBgHeight

            val highlightTextColor = ta.getColor(R.styleable.Chart_highlightTextColor, Color.WHITE)
            this.highlightTextColor = highlightTextColor

            val highlightTextSize = ta.getDimensionPixelSize(R.styleable.Chart_highlightTextSize, 28)
            this.highlightTextSize = highlightTextSize

            val isShowMaxMinValue = ta.getBoolean(R.styleable.Chart_showMaxMinValue, false)
            this.isShowMaxMinValue = isShowMaxMinValue

            val maxMinValueTextSize = ta.getDimensionPixelSize(R.styleable.Chart_maxMinValueTextSize, 28)
            this.maxMinValueTextSize = maxMinValueTextSize

            val maxMinValueTextColor = ta.getColor(R.styleable.Chart_maxMinValueTextColor, Color.WHITE)
            this.maxMinValueTextColor = maxMinValueTextColor

            val lastPriceLineColor = ta.getColor(R.styleable.Chart_lastPriceLineColor, Color.WHITE)
            this.lastPriceLineColor = lastPriceLineColor

            initAxisRenderers(ta)

            ta.recycle()
        } catch (e: Exception) {
            ta.recycle()
        }
        initChart()
    }

    private fun initAxisRenderers(ta: TypedArray) {
        axisTop.isGridLineEnable = false
        axisTop.isLabelEnable = false

        axisRenderers[TYPE_AXIS_LEFT] = AxisRenderer(this, axisLeft)
        axisRenderers[TYPE_AXIS_TOP] = AxisRenderer(this, axisTop)
        axisRenderers[TYPE_AXIS_RIGHT] = AxisRenderer(this, axisRight)
        axisRenderers[TYPE_AXIS_BOTTOM] = AxisRenderer(this, axisBottom)

        val labelTextSize = ta.getDimension(R.styleable.Chart_labelTextSize, 28f)
        val labelSeparation = ta.getDimensionPixelSize(R.styleable.Chart_labelSeparation, 10).toFloat()
        val labelTextColor = ta.getColor(R.styleable.Chart_labelTextColor, Color.GRAY)
        val gridThickness = ta.getDimension(R.styleable.Chart_gridThickness, 2f)
        val axisThickness = ta.getDimension(R.styleable.Chart_axisThickness, 2f)
        val gridColor = ta.getColor(R.styleable.Chart_gridColor, Color.GRAY)
        val axisColor = ta.getColor(R.styleable.Chart_axisColor, Color.GRAY)
        val bottomLabelHeight = ta.getDimensionPixelSize(R.styleable.Chart_bottomLabelHeight, 50)
        axisBottom.labelHeight = bottomLabelHeight

        for (axisRenderer in axisRenderers.values) {
            val axis = axisRenderer.axis
            axis.labelTextSize = labelTextSize
            axis.labelTextColor = labelTextColor
            axis.labelSeparation = labelSeparation
            axis.gridColor = gridColor
            axis.gridThickness = gridThickness
            axis.axisColor = axisColor
            axis.axisThickness = axisThickness
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
//        if (totalEntryCount > (chartData?.getTouchEntryCount() ?: 0)) return
        // Clips the next few drawing operations to the content area
        val clipRestoreCount = canvas.save()

        // 剪切ContentRect区域
        canvas.clipRect(contentRect)

        // 画背景
        if (bgColor != Color.TRANSPARENT) {
            canvas.drawColor(bgColor)
        }

        // 画水印
        if (isShowWaterMark) {
            drawWaterMark(canvas)
        }

        // 画坐标轴
        drawAxis(canvas)

        // 画网格线
        drawGridLine(canvas)

        // 画坐标轴文本 (左、右、上)
        if (isDrawLabelsInBottom) {
            drawAxisLabels(canvas)
        }

        drawChart(canvas)

        // 画线工具
        drawLineTool(canvas)

        // Removes clipping rectangle
        canvas.restoreToCount(clipRestoreCount)

        canvas.save()

        if (!isDrawLabelsInBottom) {
            drawAxisLabels(canvas)
        }

        val bottomLabelHeight = axisRenderers[TYPE_AXIS_BOTTOM]?.axis?.labelHeight ?: 0
        if (bottomLabelHeight != 0) {
            drawBottomLabels(canvas)
        }

        // 画十字光标
        drawHighlight(canvas)

        // 画区间统计
        drawRangeArea(canvas)

        canvas.restore()
    }

    override fun onSizeChanged(w: Int, h: Int, oldWidth: Int, oldHeight: Int) {
        val chartLeft = paddingLeft + if (axisLeft.isInside) 0 else axisLeft.labelWidth

        val chartRight = width - paddingRight - if (axisRight.isInside) 0 else axisRight.labelWidth

        val contentBottom = height - paddingBottom - axisBottom.labelHeight
        contentRect[chartLeft, paddingTop, chartRight] = contentBottom
        bottomRect[chartLeft, contentBottom, chartRight] = height

//        offsetPercent = axisTop.axisThickness / contentRect.height().toFloat()
        if (isShowMaxMinValue) calcMaxMinValuePercent()
    }

    private fun calcMaxMinValuePercent() {
        val rect = Rect()
        val textPaint = Paint().apply {
            textSize = maxMinValueTextSize.toFloat()
        }
        val text = "8"
        textPaint.getTextBounds(text, 0, text.length, rect)
        val height = rect.height()
        offsetPercent = height / contentRect.height().toFloat()

    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = minChartWidth + paddingLeft + (if (axisLeft.isInside) 0 else axisLeft.labelWidth) + paddingRight

        val height = minChartHeight + (if (axisBottom.isInside) 0 else axisBottom.labelHeight) + paddingBottom

        setMeasuredDimension(
            max(suggestedMinimumWidth, resolveSize(width, widthMeasureSpec)),
            max(suggestedMinimumHeight, resolveSize(height, heightMeasureSpec))
        )
    }

    fun setTypeface(tf: Typeface?) {
        axisRenderers.values.forEach { it.setTypeface(tf) }
    }

    fun getScaleY(value: Float, viewportMax: Float, viewportMin: Float): Float {
        if (viewportMax > viewportMin && viewportMax > 0) {
            return (viewportMax - value) / (viewportMax - viewportMin) * contentRect.height()
        }
        return -1f
    }

    abstract val chartData: ChartData<T>?

    abstract val renderPaint: Paint?

    abstract fun getChartAnimator(): ChartAnimator?

    /**
     * 十字光标选中
     */
    abstract fun highlightValue(highlight: Highlight?, enableCallback: Boolean = true)

    /**
     * 打开区间统计
     */
    abstract fun openRange()

    /**
     * 关闭区间统计
     */
    abstract fun closeRange()

    /**
     * 区间统计位置更新
     */
    abstract fun changeRange(startIndex: Int, endIndex: Int, touchType: Int)
}
